calcom/cal.com
# feat:  Implement A Message Queuing System

## What does this PR do?



Fixes #12556

This is an implementation of a message/task queue system implemented using [Temporal](temporal.io) (which is also an OSS project). I have added a new project (`@calcom/queue-worker`), Workflows for sending emails, and made scheduling an event to trigger the Workflow as a proof of concept.

There are some places where the value/data used could be configured to come from an environment variable but I left them hardcoded because I was running out of time to finish this hack on time 😅.

> For the email Workflows, the payload would need to be adjusted so that they're serialisable over the network. Currently, it contains a property of type `TFunction` which I think is used for translation. We would need a way to handle doing translations within the worker processes.

Loom Video: https://www.loom.com/share/b3f7d677b3cc4ec9a7f04f5eea255041?sid=39089f28-2838-48d0-a6a9-7efdcee7f1f0

## Loom Video

https://www.loom.com/share/b3f7d677b3cc4ec9a7f04f5eea255041?sid=39089f28-2838-48d0-a6a9-7efdcee7f1f0

## Requirement/Documentation



- There's an [architecture/proposal document](https://github.com/calcom/cal.com/blob/4641901720b45d6eca081e557b2d6d209bf5dfd0/apps/queue-worker/Architecture.md) which describes the solution on a high level. It also covers some info regarding the technical requirements.
- There's a [README](https://github.com/calcom/cal.com/blob/4641901720b45d6eca081e557b2d6d209bf5dfd0/apps/queue-worker/README.md) for how to run the new `queue-worker` project.

## Type of change



- [x] New feature (non-breaking change which adds functionality). Existing functionality would work as expected, only that the payload must be serialisable.

## How should this be tested?



1. Install the Temporal CLI using the command `brew install temporal`.
2. Start the Temporal dev server using `temporal server start-dev`.
3. Check out the branch and run `yarn install`.
4. Add a `.env` file to the **apps/queue-worker** project. You can copy the content of `.env.example` and make sure the `DATABASE_URL` matches the one also in the root `.env` file.
5. While in the queue-worker directory, you can start the application by running `yarn run start`. This will start the Worker process and will poll the Temporal server for new tasks.
6. Run the Cal.com web project. Go to the root directory and run `yarn run dev` to start the web application. Note: ensure you have the environment variables properly configured.
7. Open http://localhost:3000/event-types in your browser follow one of the links to schedule a meeting. This should trigger the `sendScheduledEmail` Workflow and process the event.

If you didn't set up SMTP in your environment variable, and the feature flag is configured to send email, the Task will fail to process. The failed Task will automatically be retried until it reaches the maximum retry attempt configured.

You can view the state of the Workflows via the Temporal Web UI by going to this link -> [http://localhost:8233/namespaces/default/workflows](http://localhost:8233/namespaces/default/workflows).

<img width="1417" alt="temporal-web-ui" src="https://github.com/calcom/cal.com/assets/4036118/2cff257e-3017-4903-ae19-a6482ac9f15a">

You can stop the Worker application and schedule another event as you did previously. The Task won't be processed until a Worker is available to pick it up. Now start the Worker Process (`yarn workspace @calcom/queue-worker run dev`) and see it process the Workflow Execution.

> I show most of this in my demo [video](https://www.loom.com/share/b3f7d677b3cc4ec9a7f04f5eea255041?sid=179f190d-34c7-48b6-abe3-caf4fb3a5704).

### Running the unit and integration test

There are unit and integration tests for the sendScheduledEmail Workflow currently implemented. You can run the test using `yarn workspace @calcom/queue-worker run test` command.

## Mandatory Tasks


## Checklist



- I haven't checked if new and existing unit tests pass locally with my changes
